File: src/backend/risk/riskEngine.ts
import {
  ShipmentRiskAssessment,
  RiskFactor,
  RiskAlert,
  RiskLevel,
  QuoteDecision
} from '../../types/milestone3';
import { SEEDED_RISK_ASSESSMENTS } from '../../data/milestone3Data';

const riskAssessmentsStore: Map<string, ShipmentRiskAssessment> = new Map();
const riskAlertsStore: Map<string, RiskAlert> = new Map();

// Seed initial risk assessments
Object.entries(SEEDED_RISK_ASSESSMENTS).forEach(([key, assessment]) => {
  riskAssessmentsStore.set(key, assessment);
  if (assessment.shipment_id) {
    riskAssessmentsStore.set(assessment.shipment_id, assessment);
  }
  assessment.alerts.forEach(alert => {
    riskAlertsStore.set(alert.id, alert);
  });
});

export interface RiskAssessParams {
  shipmentId: string;
  quoteId?: string;
  weatherScore?: number;
  customsScore?: number;
  routeScore?: number;
  portScore?: number;
  cargoScore?: number;
  isHazmat?: boolean;
  requiresCustomsSignoff?: boolean;
  customsSignoffCompleted?: boolean;
  customsStatus?: string;
  originPort?: string;
  destPort?: string;
  commodity?: string;
}

/**
 * Shipment Risk Engine:
 * Combines 5 weighted dimensions:
 *   Weather: 30%
 *   Customs: 25%
 *   Route:   20%
 *   Port:    15%
 *   Cargo:   10%
 *
 * overall_score =
 *   weather_score * 0.30
 * + customs_score * 0.25
 * + route_score * 0.20
 * + port_score * 0.15
 * + cargo_score * 0.10
 *
 * Thresholds:
 *   0–30: LOW
 *   31–60: MEDIUM
 *   61–80: HIGH
 *   81–100: CRITICAL
 */
export function assessShipmentCompositeRisk(params: RiskAssessParams): ShipmentRiskAssessment {
  const shipmentId = params.shipmentId || `SHP-${Math.floor(1000 + Math.random() * 9000)}`;
  const quoteId = params.quoteId || `Q-${Math.floor(1000 + Math.random() * 9000)}`;

  const weatherScore = Math.min(100, Math.max(0, params.weatherScore ?? 22));
  const customsScore = Math.min(100, Math.max(0, params.customsScore ?? (params.isHazmat ? 75 : 20)));
  const routeScore = Math.min(100, Math.max(0, params.routeScore ?? 25));
  const portScore = Math.min(100, Math.max(0, params.portScore ?? 20));
  const cargoScore = Math.min(100, Math.max(0, params.cargoScore ?? (params.isHazmat ? 80 : 15)));

  // Weighted contributions
  const weatherContrib = Number((weatherScore * 0.30).toFixed(2));
  const customsContrib = Number((customsScore * 0.25).toFixed(2));
  const routeContrib = Number((routeScore * 0.20).toFixed(2));
  const portContrib = Number((portScore * 0.15).toFixed(2));
  const cargoContrib = Number((cargoScore * 0.10).toFixed(2));

  const overallScore = Math.round(weatherContrib + customsContrib + routeContrib + portContrib + cargoContrib);

  let riskLevel: RiskLevel = 'LOW';
  if (overallScore >= 81) riskLevel = 'CRITICAL';
  else if (overallScore >= 61) riskLevel = 'HIGH';
  else if (overallScore >= 31) riskLevel = 'MEDIUM';

  const isCustomsRejected = params.customsStatus === 'REJECTED' || params.customsStatus === 'FAIL';
  const isCustomsApproved = params.customsStatus === 'APPROVED' || params.customsSignoffCompleted;
  const isCustomsHeld = params.customsStatus === 'NEEDS_DOCUMENTS' || params.customsStatus === 'NEEDS_REVIEW';

  let quoteDecision: QuoteDecision = 'APPROVED';
  let decisionRationale = 'Composite risk is within acceptable parameters. Quote issuance is approved.';

  if (isCustomsRejected || overallScore >= 81) {
    quoteDecision = 'BLOCKED';
    decisionRationale = isCustomsRejected
      ? 'Customs Compliance Officer has REJECTED the compliance check or prohibited goods were identified. Quote is BLOCKED.'
      : 'Composite Risk Score is in the CRITICAL zone (>80). Automatic and manual quotation blocked by enterprise safety policy.';
  } else if (isCustomsHeld || !isCustomsApproved || overallScore >= 61) {
    quoteDecision = 'NEEDS_REVIEW';
    decisionRationale = isCustomsHeld
      ? 'Customs documentation is incomplete or pending Customs Officer sign-off. Quote is placed on COMPLIANCE HOLD.'
      : 'Composite Risk Score is elevated (HIGH: 61-80). Requires human supervisory authorization before quote issuance.';
  }

  const assessmentId = `RSK-${Math.floor(1000 + Math.random() * 9000)}`;

  const factors: RiskFactor[] = [
    {
      id: `RF-${Math.floor(100 + Math.random() * 900)}`,
      risk_assessment_id: assessmentId,
      factor_type: 'WEATHER',
      factor_name: 'Maritime Weather & Delay Probability',
      score: weatherScore,
      weight: 0.30,
      contribution: weatherContrib,
      severity: weatherScore > 60 ? 'HIGH' : weatherScore > 30 ? 'MEDIUM' : 'LOW',
      reason: weatherScore > 60
        ? 'Severe gale/monsoon depression along corridor. High probability of vessel speed reduction.'
        : 'Favorable sailing conditions with minimal sea state resistance.',
      source: 'NOAA GFS & Copernicus Marine Radar',
      created_at: new Date().toISOString(),
    },
    {
      id: `RF-${Math.floor(100 + Math.random() * 900)}`,
      risk_assessment_id: assessmentId,
      factor_type: 'CUSTOMS',
      factor_name: 'Regulatory & Document Readiness',
      score: customsScore,
      weight: 0.25,
      contribution: customsContrib,
      severity: customsScore > 60 ? 'HIGH' : customsScore > 30 ? 'MEDIUM' : 'LOW',
      reason: customsScore > 60
        ? 'Pending mandatory compliance certificates or regulatory permits (MSDS/FSSAI/CECA).'
        : 'All core commercial trade documents verified on ICEGATE.',
      source: 'Customs Intelligence Agent & CBIC Schedule',
      created_at: new Date().toISOString(),
    },
    {
      id: `RF-${Math.floor(100 + Math.random() * 900)}`,
      risk_assessment_id: assessmentId,
      factor_type: 'ROUTE',
      factor_name: 'Corridor & Chokepoint Safety',
      score: routeScore,
      weight: 0.20,
      contribution: routeContrib,
      severity: routeScore > 60 ? 'HIGH' : routeScore > 30 ? 'MEDIUM' : 'LOW',
      reason: routeScore > 60
        ? 'High traffic density in maritime chokepoints or potential coastal deviation.'
        : 'Standard direct navigation lane with high historical reliability.',
      source: 'IMU Marine Route Graph',
      created_at: new Date().toISOString(),
    },
    {
      id: `RF-${Math.floor(100 + Math.random() * 900)}`,
      risk_assessment_id: assessmentId,
      factor_type: 'PORT',
      factor_name: 'Port Congestion & Terminal Dwell',
      score: portScore,
      weight: 0.15,
      contribution: portContrib,
      severity: portScore > 60 ? 'HIGH' : portScore > 30 ? 'MEDIUM' : 'LOW',
      reason: portScore > 60
        ? 'Elevated container dwell and berth queue at origin/destination hubs.'
        : 'Origin and destination terminal turnarounds are within optimal benchmarks (<2 days).',
      source: 'Global AIS & Port Master Dwell Analytics',
      created_at: new Date().toISOString(),
    },
    {
      id: `RF-${Math.floor(100 + Math.random() * 900)}`,
      risk_assessment_id: assessmentId,
      factor_type: 'CARGO',
      factor_name: 'Commodity Risk & Hazmat Classification',
      score: cargoScore,
      weight: 0.10,
      contribution: cargoContrib,
      severity: cargoScore > 60 ? 'HIGH' : cargoScore > 30 ? 'MEDIUM' : 'LOW',
      reason: cargoScore > 60
        ? 'Dangerous goods / temperature-controlled cargo requiring specialized container handling.'
        : 'Standard dry general cargo with low damage/theft susceptibility.',
      source: 'IMDG Hazardous Cargo Code & Master Data',
      created_at: new Date().toISOString(),
    }
  ];

  const alerts: RiskAlert[] = [];
  if (quoteDecision === 'BLOCKED' || quoteDecision === 'NEEDS_REVIEW') {
    const alert: RiskAlert = {
      id: `ALT-RSK-${Math.floor(1000 + Math.random() * 9000)}`,
      shipment_id: shipmentId,
      quote_id: quoteId,
      risk_assessment_id: assessmentId,
      alert_type: quoteDecision === 'BLOCKED' ? 'QUOTE_BLOCKED_CRITICAL' : 'QUOTE_HOLD_COMPLIANCE',
      severity: quoteDecision === 'BLOCKED' ? 'CRITICAL' : 'WARNING',
      title: quoteDecision === 'BLOCKED' ? 'Critical Shipment Risk Triggered' : 'Compliance & Risk Review Required',
      message: decisionRationale,
      source: 'Shipment Risk Engine',
      status: 'ACTIVE',
      created_at: new Date().toISOString(),
    };
    alerts.push(alert);
    riskAlertsStore.set(alert.id, alert);
  }

  const assessment: ShipmentRiskAssessment = {
    id: assessmentId,
    shipment_id: shipmentId,
    quote_id: quoteId,
    weather_score: weatherScore,
    customs_score: customsScore,
    route_score: routeScore,
    port_score: portScore,
    cargo_score: cargoScore,
    overall_score: overallScore,
    risk_level: riskLevel,
    confidence_score: 0.95,
    explanation: `Composite risk score is ${overallScore} (${riskLevel}). Weather contrib: ${weatherContrib}, Customs contrib: ${customsContrib}, Route contrib: ${routeContrib}, Port contrib: ${portContrib}, Cargo contrib: ${cargoContrib}.`,
    quote_decision: quoteDecision,
    requires_customs_signoff: params.requiresCustomsSignoff ?? true,
    customs_signoff_completed: isCustomsApproved,
    decision_rationale: decisionRationale,
    factors: factors,
    alerts: alerts,
    assessed_at: new Date().toISOString(),
    model_version: 'RiskEngine-v3.2-Hybrid',
    created_at: new Date().toISOString(),
    updated_at: new Date().toISOString(),
  };

  riskAssessmentsStore.set(assessmentId, assessment);
  riskAssessmentsStore.set(shipmentId, assessment);
  return assessment;
}

export function getShipmentRisk(shipmentId: string): ShipmentRiskAssessment | undefined {
  return riskAssessmentsStore.get(shipmentId);
}

export function getAllRiskAlerts(): RiskAlert[] {
  return Array.from(riskAlertsStore.values());
}

export function acknowledgeRiskAlert(alertId: string, ackUser: string = 'User'): boolean {
  const alert = riskAlertsStore.get(alertId);
  if (!alert) return false;
  alert.status = 'ACKNOWLEDGED';
  alert.acknowledged_by = ackUser;
  alert.acknowledged_at = new Date().toISOString();
  return true;
}

File: src/backend/services/risk_engine.py
"""
Milestone 3 (M3) — Composite Shipment Risk Engine Service
Fuses 5 dimensions: Weather (30%), Customs (25%), Route (20%), Port (15%), and Cargo (10%).
"""

from datetime import datetime, timezone
from typing import List, Optional
from pydantic import BaseModel, Field

from src.backend.models.m3_models import (
    ShipmentRiskAssessment,
    RiskFactor,
    RiskAlert,
    RiskFactorType,
    RiskLevel,
    AlertSeverity,
    AlertStatus,
    QuoteDecision,
    utc_now,
)


class RiskAssessRequest(BaseModel):
    shipment_id: Optional[str] = Field(None, description="Linked shipment ID")
    quote_id: Optional[str] = Field(None, description="Linked quote ID")
    weather_score: Optional[float] = Field(None, ge=0.0, le=100.0, description="0-100 Weather risk score")
    customs_score: Optional[float] = Field(None, ge=0.0, le=100.0, description="0-100 Customs readiness risk score")
    route_score: Optional[float] = Field(None, ge=0.0, le=100.0, description="0-100 Geopolitical & chokepoint score")
    port_score: Optional[float] = Field(None, ge=0.0, le=100.0, description="0-100 Port congestion score")
    cargo_score: Optional[float] = Field(None, ge=0.0, le=100.0, description="0-100 Commodity hazmat/volatility score")
    is_hazmat: bool = Field(default=False)
    customs_status: Optional[str] = Field(default="NEEDS_REVIEW")
    requires_customs_signoff: bool = Field(default=True)
    customs_signoff_completed: bool = Field(default=False)
    origin_port: Optional[str] = Field(default="INMAA")
    destination_port: Optional[str] = Field(default="SGSIN")
    commodity: Optional[str] = Field(default="Standard Cargo")


class RiskEngineService:
    """
    Composite 5-Pillar Shipment Risk Engine:
    Weighted Scoring Formula:
      overall_score = (weather_score * 0.30)
                    + (customs_score * 0.25)
                    + (route_score   * 0.20)
                    + (port_score    * 0.15)
                    + (cargo_score   * 0.10)
    """

    # In-memory backup cache for instant retrieval
    _cache: dict[str, ShipmentRiskAssessment] = {}

    @classmethod
    async def assess_risk(cls, req: RiskAssessRequest) -> ShipmentRiskAssessment:
        shipment_id = req.shipment_id or f"SHP-{int(datetime.now().timestamp())}"
        quote_id = req.quote_id or f"Q-{int(datetime.now().timestamp())}"

        # 1. Normalize sub-scores (0-100)
        weather_s = min(100.0, max(0.0, req.weather_score if req.weather_score is not None else 24.0))
        customs_s = min(100.0, max(0.0, req.customs_score if req.customs_score is not None else (75.0 if req.is_hazmat else 22.0)))
        route_s = min(100.0, max(0.0, req.route_score if req.route_score is not None else 25.0))
        port_s = min(100.0, max(0.0, req.port_score if req.port_score is not None else 20.0))
        cargo_s = min(100.0, max(0.0, req.cargo_score if req.cargo_score is not None else (80.0 if req.is_hazmat else 15.0)))

        # 2. Compute weighted contributions
        weather_contrib = round(weather_s * 0.30, 2)
        customs_contrib = round(customs_s * 0.25, 2)
        route_contrib = round(route_s * 0.20, 2)
        port_contrib = round(port_s * 0.15, 2)
        cargo_contrib = round(cargo_s * 0.10, 2)

        overall_score = round(weather_contrib + customs_contrib + route_contrib + port_contrib + cargo_contrib, 1)

        # 3. Determine Risk Tier
        if overall_score >= 81.0:
            risk_level = RiskLevel.CRITICAL
        elif overall_score >= 61.0:
            risk_level = RiskLevel.HIGH
        elif overall_score >= 31.0:
            risk_level = RiskLevel.MEDIUM
        else:
            risk_level = RiskLevel.LOW

        # 4. Evaluate automated quotation gating
        is_customs_rejected = req.customs_status in ["REJECTED", "FAIL"]
        is_customs_approved = req.customs_status == "APPROVED" or req.customs_signoff_completed
        is_customs_held = req.customs_status in ["NEEDS_DOCUMENTS", "NEEDS_REVIEW", "PENDING"]

        if is_customs_rejected or overall_score >= 81.0:
            quote_decision = QuoteDecision.BLOCKED
            decision_rationale = (
                "Customs Compliance REJECTED or Prohibited Commodity Identified."
                if is_customs_rejected
                else "Composite risk score exceeds Enterprise Safety Threshold (CRITICAL >80). Instant automated quoting blocked."
            )
        elif is_customs_held or not is_customs_approved or overall_score >= 61.0:
            quote_decision = QuoteDecision.NEEDS_REVIEW
            decision_rationale = (
                "Mandatory trade permits or customs sign-off pending. Quote placed on Compliance Hold."
                if is_customs_held
                else "Elevated composite risk (HIGH: 61-80). Requires human trade officer approval prior to release."
            )
        else:
            quote_decision = QuoteDecision.APPROVED
            decision_rationale = "Composite risk is within standard parameters. Automated quotation is authorized."

        # 5. Build factor-level explainability breakdown
        factors: List[RiskFactor] = [
            RiskFactor(
                id="RF-WTR-01",
                factor_type=RiskFactorType.WEATHER,
                factor_name="Marine Weather & Delay Probability",
                score=weather_s,
                weight=0.30,
                contribution=weather_contrib,
                severity=RiskLevel.HIGH if weather_s > 60 else (RiskLevel.MEDIUM if weather_s > 30 else RiskLevel.LOW),
                reason="Severe gale/swell alert along transit trench." if weather_s > 60 else "Favorable sea state and low squall probability.",
                source="NOAA WaveWatch III & Copernicus Radar",
            ),
            RiskFactor(
                id="RF-CUST-02",
                factor_type=RiskFactorType.CUSTOMS,
                factor_name="Customs Tariff & Document Readiness",
                score=customs_s,
                weight=0.25,
                contribution=customs_contrib,
                severity=RiskLevel.HIGH if customs_s > 60 else (RiskLevel.MEDIUM if customs_s > 30 else RiskLevel.LOW),
                reason="Pending required import permits or ICEGATE declaration verification." if customs_s > 60 else "Statutory documentation verified.",
                source="Customs Agent RAG & CBIC Schedule",
            ),
            RiskFactor(
                id="RF-RTE-03",
                factor_type=RiskFactorType.ROUTE,
                factor_name="Corridor & Chokepoint Safety",
                score=route_s,
                weight=0.20,
                contribution=route_contrib,
                severity=RiskLevel.HIGH if route_s > 60 else (RiskLevel.MEDIUM if route_s > 30 else RiskLevel.LOW),
                reason="Transit via elevated geopolitical traffic lane." if route_s > 60 else "Standard commercial sea lane with high reliability.",
                source="IMU Marine Route Graph",
            ),
            RiskFactor(
                id="RF-PRT-04",
                factor_type=RiskFactorType.PORT,
                factor_name="Port Congestion & Terminal Dwell",
                score=port_s,
                weight=0.15,
                contribution=port_contrib,
                severity=RiskLevel.HIGH if port_s > 60 else (RiskLevel.MEDIUM if port_s > 30 else RiskLevel.LOW),
                reason="Elevated container berth queue and yard congestion at destination." if port_s > 60 else "Optimal turnaround times (<48h).",
                source="Global Port Master & AIS Telemetry",
            ),
            RiskFactor(
                id="RF-CRG-05",
                factor_type=RiskFactorType.CARGO,
                factor_name="Commodity & Dangerous Goods Classification",
                score=cargo_s,
                weight=0.10,
                contribution=cargo_contrib,
                severity=RiskLevel.HIGH if cargo_s > 60 else (RiskLevel.MEDIUM if cargo_s > 30 else RiskLevel.LOW),
                reason="Hazardous materials (IMDG) or temperature-sensitive cargo." if cargo_s > 60 else "Standard dry containerized freight.",
                source="IMDG Dangerous Goods Code & Master Data",
            ),
        ]

        # 6. Generate active alerts if risk is elevated
        alerts: List[RiskAlert] = []
        if quote_decision in [QuoteDecision.BLOCKED, QuoteDecision.NEEDS_REVIEW]:
            alert = RiskAlert(
                shipment_id=shipment_id,
                quote_id=quote_id,
                risk_assessment_id="RSK-PENDING",
                alert_type="QUOTE_BLOCKED_CRITICAL" if quote_decision == QuoteDecision.BLOCKED else "QUOTE_HOLD_COMPLIANCE",
                severity=AlertSeverity.CRITICAL if quote_decision == QuoteDecision.BLOCKED else AlertSeverity.WARNING,
                title="Critical Shipment Risk Triggered" if quote_decision == QuoteDecision.BLOCKED else "Customs & Risk Review Required",
                message=decision_rationale,
                source="Shipment Risk Engine v3.4",
                status=AlertStatus.ACTIVE,
            )
            alerts.append(alert)

        # 7. Create and persist assessment
        assessment = ShipmentRiskAssessment(
            shipment_id=shipment_id,
            quote_id=quote_id,
            weather_score=weather_s,
            customs_score=customs_s,
            route_score=route_s,
            port_score=port_s,
            cargo_score=cargo_s,
            overall_score=overall_score,
            risk_level=risk_level,
            confidence_score=0.96,
            explanation=(
                f"Composite score is {overall_score}/100 ({risk_level.value}). "
                f"Weather: {weather_contrib} pts, Customs: {customs_contrib} pts, "
                f"Route: {route_contrib} pts, Port: {port_contrib} pts, Cargo: {cargo_contrib} pts."
            ),
            quote_decision=quote_decision,
            requires_customs_signoff=req.requires_customs_signoff,
            customs_signoff_completed=is_customs_approved,
            decision_rationale=decision_rationale,
            factors=factors,
            alerts=alerts,
            assessed_at=utc_now(),
            model_version="RiskEngine-v3.4-Fused",
        )

        try:
            await assessment.save()
        except Exception:
            pass

        cls._cache[shipment_id] = assessment
        return assessment

    @classmethod
    async def get_assessment_by_shipment(cls, shipment_id: str) -> Optional[ShipmentRiskAssessment]:
        """Fetch existing risk assessment or synthesize a baseline."""
        try:
            doc = await ShipmentRiskAssessment.find_one(ShipmentRiskAssessment.shipment_id == shipment_id)
            if doc:
                return doc
        except Exception:
            pass

        if shipment_id in cls._cache:
            return cls._cache[shipment_id]

        # Generate on the fly if not found
        req = RiskAssessRequest(shipment_id=shipment_id)
        return await cls.assess_risk(req)
